### ISSUE TITLE
Use within_bounding_box API endpoint and getMapBoundaries function for loading map markers that are within viewport

### ISSUE THREAD

**Issue by RyanTG** [pinballmap/pbm-react#449]
Rather than a radius (which results in lots of markers loading off-screen to the right and left, could we use the `getMapBoundaries` function (from react native maps) to only load markers that are in the visible viewport (with maybe a small buffer)?

**Comment by RyanTG** [comment]
Oh, maybe not, since we use the closest_by_lat_lon.json endpoint and that is what is pulling everything within a radius?

**Comment by RyanTG** [comment]
Added a `within_bounding_box` API endpoint. https://github.com/pinballmap/pbm/commit/b1656a64595ba094d5349b15199bec07337c5667

Not yet sure if this will work perfectly with `getMapBoundaries`. 

`within_bounding_box` looks for this array: [swlat, swlon, nelat, nelon]

However, `getMapBoundies` returns them in the wrong order: `Promise<{northEast: LatLng, southWest: LatLng}>`

But at this point I'm assuming we can grab them from getMapBoundaries and swap them around.

Example url: `http://localhost:3000/api/v1/locations/within_bounding_box.json?swlat=34.16393434457936;swlon=-118.55308182989683;nelat=34.27249281510788;nelon=-118.39944560363433`

Right now I have each lat and lon as a separate param:

```
param :swlat, String, 'SW_Latitude', required: true
param :swlon, String, 'SW_Longitude', required: true
param :nelat, String, 'NE_Latitude', required: true
param :nelon, String, 'NE_Longitude', required: true

bounds = [ params[:swlat], params[:swlon], params[:nelat], params[:nelon] ]
locations_within = apply_scopes(Location).includes(:machines).within_bounding_box(bounds)
```

But if `getMapBoundaries` is returning us southWest and northEast, and we can't access the individual lat/lons of both of those (to pass them as four separate params), then we'll need to rethink this.

**Comment by RyanTG** [comment]
Once this is added, then I would like to review the "zoom limits" that we are using, with the goal of letting users zoom much further out, especially when filters are applied.

**Comment by bpoore** [comment]
Some notes for consideration and just to keep track of things :
* `getMapBoundaries` gives us the coordinates we need to use the `within_bounding_box` endpoint- so that's great. Unfortunately, it's not a simple clean slide-in / replacement for how we're currently fetching locations. For instance, we immediately run into needing to change our mental model when the app is even first loaded- we currently place the map at a lat/lon with latDelta/lonDelta defaulted to 0.1- in reality the latDelta/lonDelta are not actually rendered at 0.1, so we don't have the "bounding box" coordinates until the map loads- which we can handle by not making the request for machines until the map renders (shouldn't be a noticeable delay)
* I was initially perplexed regarding how we want to handle requests for region and city searches. Looks like for searching by city, we'll be in a better spot overall because we have the coordinates for the cities, so we'll be able to plug in the [max coords](https://github.com/pinballmap/pbm-react/blob/master/app/components/Search.js#L129) we're already determining. For region search, we're using the effective radius from the api to determine an approximate [lat/lon delta](https://github.com/pinballmap/pbm-react/blob/master/app/actions/locations_actions.js#L126) for the map to zoom to. I think we can treat this like the case where the map originally loads where we would get the coordinates after the map renders? Still thinking it through for that one. Hoping to not have several ways of keeping track what the coordinates of the map currently are / what the zoom should be.
* How are we going to deal with the concept of maxZoom? Are we still going to approximate the amount of viewable distance on the screen?

I think it's possible to get this right, but it will require a fair amount of time and effort. If we feel adjusting to use this new endpoint really provides a big win for users, I'm open to working through this, but saying now- these are not trivial changes and this will require a lot of regression testing.

**Comment by RyanTG** [comment]
For region and city, could we use one of RNMaps’ “fit” functions - Like fitToElements or fitToSuppliedMarkers? I think we explored this previously and it didn’t work for us. For city, I don’t think the API gives us a list of location IDs based on city, state.

Yeah, this does seem like a lot to consider. When I get back I’ll refamiliarize myself with how we’re doing stuff.

I do think overall this would be a big win. We could potentially greatly expand the max zoom, which is something we get a lot of comments about.

**Comment by RyanTG** [comment]
After more testing of the bounding box that is now merged into master, I am seeing that when I first open the app, it is consistently not including locations that are south of my location (just above the tab bar). Clicking refresh causes them to show up. I also see this if I move away from my location, and then click the "my location" button (same locations south of me do not show up).

Are you seeing this? I tried both on staging and prod.

**Comment by bpoore** [comment]
https://github.com/pinballmap/pbm-react/commit/b4a09c11e0a303e72d0a7a06bb11ad98773c1dba

**Comment by RyanTG** [comment]
Let me know if I'm missing something here.

I added two more params to the bounding box endpoint. Basically, I didn't want to mess with `by_machine_id`, so instead I added two new params for explicitly requesting either the single version or all versions in the machine group. I was thinking we could use these _instead of_ `by_machine_id` because I would like to have different max zoom thresholds for each.

- by_machine_single_id (takes the machine ID)
- by_machine_group_id (takes the machine _group_ ID)

When a machine is not part of a group, I think it should use by_machine_single_id (so that we can give it the same max zoom threshold as when a user specifically filters "single version").

Does this make sense?
